app/models/info_request.rb

```diff
@@ -185,7 +185,7 @@ class << self
   after_destroy :update_counter_cache
   after_update :reindex_some_request_events
   before_destroy :expire
-  before_validation :compute_idhash
+  before_validation :compute_idhash, :set_law_used
```

Comment:
It would make more sense conceptually to use an [association callback](https://guides.rubyonrails.org/v5.2/association_basics.html#association-callbacks), since this is closer to what we _mean_ to happen – when we update the associated `PublicBody`, make sure the `InfoRequest` is being made under the correct `Legislation`.

I think this would resolve the comments on the previous two commits. We'd be able to set a default (since we know it will get updated correctly when the body is assigned) and we can retain the existing validations that mirror the database constraints, since we no longer need to support a `nil` `law_used` attribute.

The thing to consider here is what should happen if a request is moved from, say, an FOI-able body to an EIR-only body. Something like the following might work?

```ruby
before_add: [:check_law_used_supported, :update_law_used]

def check_law_used_supported(public_body)
  return true if new_record?
  raise SomeError unless public_body.covered_by?(legislation)
end

def update_law_used(public_body)
  self.law_used = public_body.legislation
end
```

I don't think I'd worry too much about handling the exception in the admin UI at this point – if it happens a lot then we can ticket up and do something.

Reply:
Rails doesn't support association callbacks on `belongs_to` associations (see https://github.com/rails/rails/issues/586). We could override the `#public_body=` if you prefer but a before validation callback seemed simpler.

Currently when moving a request to a different public body we update the `#law_used` https://github.com/mysociety/alaveteli/blob/b1d999103889f6754d78d49e92cfad41e3302017/app/models/info_request.rb#L1620-L1622

This PR doesn't change this behaviour as its not really within scope but I don't think we should be updating `#law_used` as once the request has been created the application wouldn't know if the user wrote a FOI or EIR request. Its entirely possible a user would attempt to send a FOI to an EIR-only body.

We could even lose the correct law used when moving, suppose a EIR request was written to a FOI&EIR public body, the request moved to another FOI&EIR body - I think the law used would now be set to `foi`. If you agree I'll open a ticket for this -  the solution could when look a bit like your `#check_law_used_supported`.